### PR TITLE
Song select filtering

### PR DESCRIPTION
**Pull request by ddevault**
Includes basic sorting+grouping UI, implementation to come separately.

### CODE REVIEW

File: osu.Desktop/Program.cs [hunk 1]
```diff
@@ -25,17 +26,22 @@ public static class Program
...
-
+                    Directory.SetCurrentDirectory(cwd);
                     foreach (var file in args)
-                        if (!importer.Import(file).Wait(1000))
+                    {
+                        Console.WriteLine(@"Importing {0}", file);
+                        if (!importer.Import(Path.GetFullPath(file)).Wait(3000))
                             throw new TimeoutException(@"IPC took too long to send");
-                    Console.WriteLine(@"Sent import requests to running instance");
+                        Thread.Sleep(500);
```
Comment (peppy) [inline osu.Desktop/Program.cs]: why 500ms?

File: osu.Desktop/Program.cs [hunk 2]
```diff
@@ -25,17 +26,22 @@ public static class Program
...
+            
+            var cwd = Directory.GetCurrentDirectory();
 
             using (DesktopGameHost host = Host.GetSuitableHost(@"osu", true))
             {
                 if (!host.IsPrimaryInstance)
                 {
                     var importer = new BeatmapImporter(host);
-
+                    Directory.SetCurrentDirectory(cwd);
```
Comment (peppy) [inline osu.Desktop/Program.cs]: what's this for?

File: osu.Desktop/Program.cs [hunk 3]
```diff
@@ -25,17 +26,22 @@ public static class Program
         public static int Main(string[] args)
         {
             LegacyFilesystemReader.Register();
+            
+            var cwd = Directory.GetCurrentDirectory();
```
Comment (peppy) [inline osu.Desktop/Program.cs]: `Environment.CurrentDirectory` may be a better fit? not sure what you're using this for though. needs comments since it's not obvious.

File: osu.Game/Beatmaps/Drawables/BeatmapGroup.cs [hunk 4]
```diff
@@ -1,42 +1,46 @@
...
 
-        public List<BeatmapPanel> BeatmapPanels;
-
-        public BeatmapGroupState State
-        {
-            get { return state; }
-            set
-            {
-                state = value;
+        public bool Hidden;
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/BeatmapGroup.cs]: not sure i like this being a public bool with no actual logic behind it. changing this value should either not be allowed or should actually perform an action. consider making it private if it's just storing a local state.

File: osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs [hunk 5]
```diff
@@ -56,6 +55,8 @@ public WorkingBeatmap Beatmap
         public BackgroundModeBeatmap(WorkingBeatmap beatmap)
         {
             Beatmap = beatmap;
+            if (beatmap == null)
+                Add(background = new Background(@"Backgrounds/bg1"));
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: move this logic to line29 and combine with the rest of the logic that is adding the background (and apply the same blur logic etc. it should behave exactly the same.

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 6]
```diff
@@ -248,6 +251,13 @@ protected override void Update()
             }
         }
 
+        public void InvalidateVisible()
```
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: what does this even mean? nothing is being invalidated here.

File: osu.Game/Screens/Select/FilterControl.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Input;
+using osu.Game.Graphics;
+
+namespace osu.Game.Screens.Select
+{
+    public class FilterControl : Container
+    {
+        public enum SortMode
```
Comment (peppy) [inline osu.Game/Screens/Select/FilterControl.cs]: i prefer enums going at the end of the file

File: osu.Game/Screens/Select/FilterControl.cs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            Collections,
+            Favorites,
+            MyMaps,
+            RankedStatus,
+            RecentlyPlayed
+        }
+    
+        public Action FilterChanged;
+
+        public string Search { get; private set; } = string.Empty;
```
Comment (peppy) [inline osu.Game/Screens/Select/FilterControl.cs]: any reason this needs storage as opposed to just linking directly to searchTextBox?

ie. 
`public string Search => searchTextBox.Text`

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 9]
```diff
@@ -174,9 +187,55 @@ private void start()
             Task.Factory.StartNew(() => addBeatmapSets(game, initialAddSetsTask.Token), initialAddSetsTask.Token);
         }
 
+        private ScheduledDelegate filterTask;
+
+        private void filterChanged()
+        {
+            if (filterTask != null)
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: `filterTask?.Cancel()`

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 10]
```diff
@@ -174,9 +187,55 @@ private void start()
...
+            filterTask = Scheduler.AddDelayed(() =>
+            {
+                filterTask = null;
+                var search = filter.Search;
+                BeatmapGroup newSelection = null;
+                bool changed = false;
+                foreach (var beatmapGroup in carousel)
+                {
+                    var set = beatmapGroup.BeatmapSet;
+                    if (set == null)
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: is this ever a thing?

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 11]
```diff
@@ -174,9 +187,55 @@ private void start()
...
+                        continue;
+                    bool match = string.IsNullOrEmpty(search)
+                        || (set.Metadata.Artist ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.ArtistUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.Title ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.TitleUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1;
+                    if (match)
+                    {
+                        changed = changed && !beatmapGroup.Hidden;
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: `change &= !beatmapGroup.Hidden`

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 12]
```diff
@@ -174,9 +187,55 @@ private void start()
...
+                        beatmapGroup.Hidden = false;
+                        beatmapGroup.Header.Alpha = 1;
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: ugly. make Hidden a property and actually do this inside the beatmapGroup.

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 13]
```diff
@@ -174,9 +187,55 @@ private void start()
...
+                        if (newSelection == null || beatmapGroup.BeatmapSet.OnlineBeatmapSetID == Beatmap.BeatmapSetInfo.OnlineBeatmapSetID)
+                            newSelection = beatmapGroup;
+                    }
+                    else
+                    {
+                        changed = changed && beatmapGroup.Hidden;
+                        beatmapGroup.Hidden = true;
+                        beatmapGroup.Header.Alpha = 0;
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: ditto

**Comment by peppy** [comment]
- Escape key doesn't work at song select any more.
- Consider making the Hidden thing a state of `BeatmapGroupState` rather than this `bool Hidden` poop.

File: osu.Desktop/Program.cs (hunk 1 above)
Comment (SirCmpwn) [inline osu.Desktop/Program.cs]: Arbitrary number to slow it down, since the server doesn't tell us when it's done processing an import. We could refactor this but should be done separately imo.

File: osu.Desktop/Program.cs (hunk 2 above)
Comment (SirCmpwn) [inline osu.Desktop/Program.cs]: DesktopGameHost changes the cwd to the path of the binary. This changes it back so relative paths make sense when passed at the command line.

File: osu.Desktop/Program.cs (hunk 3 above)
Comment (SirCmpwn) [inline osu.Desktop/Program.cs]: Will add comments.

File: osu.Game/Beatmaps/Drawables/BeatmapGroup.cs (hunk 4 above)
Comment (SirCmpwn) [inline osu.Game/Beatmaps/Drawables/BeatmapGroup.cs]: It tells the CarouselContainer how to treat it, primarily.

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 6 above)
Comment (SirCmpwn) [inline osu.Game/Screens/Select/CarouselContainer.cs]: Not invalidate in the sense of Drawable.Invalidate, but it says "the Y positions of panels is now wrong, the Hidden state of some of them has changed"

File: osu.Game/Screens/Select/PlaySongSelect.cs (hunk 11 above)
Comment (SirCmpwn) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: &&= doesn't exist in C#. Does &= do the same thing?

**Comment by SirCmpwn** [comment]
>Consider making the Hidden thing a state of BeatmapGroupState rather than this bool Hidden poop.

I did consider that, but it complicates things, imo a bit too much.

File: osu.Game/Screens/Select/PlaySongSelect.cs (hunk 10 above)
Comment (SirCmpwn) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: Just coding defensively. Will remove it.

File: osu.Game/Screens/Select/PlaySongSelect.cs (hunk 11 above)
Comment (Tom94) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: Yes. IIRC in C# && is literally &, only with short-circuiting.

File: osu.Game/Screens/Select/SearchTextBox.cs [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+        }
+        
+        protected override void Update()
+        {
+            if (GrabFocus && !HasFocus && IsVisible)
+            {
+                var inputManager = Parent;
```
Comment (peppy) [inline osu.Game/Screens/Select/SearchTextBox.cs]: i wouldn't bother adding this parent check for now. when it becomes an issue we should fix it in InputManager's Focused handling, not locally here.

**Comment by peppy** [comment]
- escape still doesn't work
- selecting any difficulty that isn't the first in a group then entering play mode causes weird shit to happen

please test stuff yourself :(

**Comment by SirCmpwn** [comment]
>escape still doesn't work

Sorry. I did test this, but not after the last set of changes.

>selecting any difficulty that isn't the first in a group then entering play mode causes weird shit to happen

Cannot reproduce.

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 15]
```diff
@@ -189,9 +189,13 @@ private void start()
         }
 
         private ScheduledDelegate filterTask;
+        private string previousFilter = null;
 
         private void filterChanged()
         {
+            if (previousFilter == filter.Search)
```
Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: did you mean to use this?

image

Comment (SirCmpwn) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: No, but I forgot to do the other thing I meant to do that justified that decision.

File: osu.Desktop/Program.cs (hunk 1 above)
Comment (peppy) [inline osu.Desktop/Program.cs]: i really don't like arbitrary sleeps. can we just leave this out for now?

File: osu.Game/Beatmaps/Drawables/BeatmapGroup.cs [hunk 16]
```diff
@@ -28,18 +28,21 @@ class BeatmapGroup : IStateful<BeatmapGroupState>
         public BeatmapSetHeader Header;
 
         private BeatmapGroupState state;
-
+
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/BeatmapGroup.cs]: trim

File: osu.Desktop/Program.cs (hunk 1 above)
Comment (SirCmpwn) [inline osu.Desktop/Program.cs]: Adding multiple beatmaps at once doesn't work if we do, which makes it pretty annoying to wipe and refill your database.

Comment (peppy) [inline osu.Desktop/Program.cs]: i've added 100 at a time with no issues until now?

Comment (SirCmpwn) [inline osu.Desktop/Program.cs]: Alright, w/e. I'll just keep it on my own builds.

File: osu.Game/Beatmaps/Drawables/BeatmapGroup.cs [hunk 17]
```diff
@@ -48,14 +51,20 @@ public BeatmapGroupState State
                             SelectedPanel.State = PanelSelectedState.Selected;
                         break;
                     case BeatmapGroupState.Collapsed:
+                    case BeatmapGroupState.Hidden:
+                        if (state == BeatmapGroupState.Hidden && state != value)
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/BeatmapGroup.cs]: `Header.Alpha = state == BeatmapGroupState.Hidden ? 0 : 1;`
put this outside the switch and remove all the conditionals you added inside it.

File: osu.Game/Screens/Play/Player.cs [hunk 18]
```diff
@@ -167,9 +167,6 @@ protected override void OnEntering(GameMode last)
         {
             base.OnEntering(last);
 
-            (Background as BackgroundModeBeatmap)?.BlurTo(Vector2.Zero, 1000);
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: why is this removed? the background no longer unblurs

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 19]
```diff
@@ -133,10 +134,10 @@ private float computeYPositions()
                         panel.MoveToX(-50, 500, EasingTypes.OutExpo);
 
                         //on first display we want to begin hidden under our group's header.
-                        if (panel.Alpha == 0)
+                        if (panel.Alpha == 0 && group.State != BeatmapGroupState.Hidden)
```
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: guaranteed to be true

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 20]
```diff
@@ -133,10 +134,10 @@ private float computeYPositions()
                         panel.MoveToX(-50, 500, EasingTypes.OutExpo);
 
                         //on first display we want to begin hidden under our group's header.
-                        if (panel.Alpha == 0)
+                        if (panel.Alpha == 0 && group.State != BeatmapGroupState.Hidden)
                             panel.MoveToY(headerY);
 
-                        movePanel(panel, true, ref currentY);
+                        movePanel(panel, group.State != BeatmapGroupState.Hidden, ref currentY);
```
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: ditto

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 21]
```diff
@@ -300,5 +310,15 @@ public void SelectRandom()
                 return;
             SelectGroup(group, panel);
         }
+
+        public IEnumerator<BeatmapGroup> GetEnumerator()
```
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: public IEnumerator<BeatmapGroup> GetEnumerator() => groups.GetEnumerator();

File: osu.Game/Screens/Select/CarouselContainer.cs [hunk 22]
```diff
@@ -300,5 +310,15 @@ public void SelectRandom()
                 return;
             SelectGroup(group, panel);
         }
+
+        public IEnumerator<BeatmapGroup> GetEnumerator()
+        {
+            return groups.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
```
Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

File: osu.Game/Screens/Select/CarouselContainer.cs (hunk 19 above)
Comment (SirCmpwn) [inline osu.Game/Screens/Select/CarouselContainer.cs]: No, during animations Alpha could be 0.

Comment (peppy) [inline osu.Game/Screens/Select/CarouselContainer.cs]: read this again

**Comment by peppy** [comment]
Posted fixes for remaining issues at https://github.com/SirCmpwn/osu/pull/1.

File: osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs [hunk 23]
```diff
@@ -22,17 +22,19 @@ public WorkingBeatmap Beatmap
...
+                if (beatmap == value && beatmap != null)
                     return;
-
                 beatmap = value;
 
                 Schedule(() =>
                 {
-                    Background newBackground = new BeatmapBackground(beatmap);
+                    Background newBackground;
+                    if (beatmap == null)
```
Comment (Tom94) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: I am against allowing beatmap == null. We should avoid the whole class of problems assiciated with null by having a default built-in beatmap. Less corner cases (if statements for things like null) means cleaner and more readable code.

Comment (SirCmpwn) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: :-1: I would find that very annoying. A whole class of problems is really just 2 or 3 problems, which have all been addressed as of this PR.

Comment (Tom94) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: What I mean with "whole class of problems" is that _every single time a beatmap is used_ you need a special treatment for the null case. Every. Single. Time. In the old code base those were literally hundreds of times.

What do you think is more annoying than that about having a default beatmap that avoids all this special treatment of a null beatmap?

Comment (SirCmpwn) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: But only on this screen. You can't get to the player if there's no song for you to select here. I'm not suggesting we allow every screen to have a null beatmap, it makes sense of course for Player to require a non-null beatmap for example.

Comment (SirCmpwn) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: Unless you want to download, process, and import syncronously (blocking the UI for several seconds average case), there will be some time when the UI is interactive and the beatmap database is empty, and thus the user can find their way over to screens where we need to handle that.

Comment (Tom94) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: You can still prohibit things or add special behavior by checking `beatmap == Default`. In the majority of cases you don't need that, though. The beatmap could have the osu! theme as song, and would be built-in into the executable anyway, so you also avoid the logic needed to play the theme song when no beatmap is selected (which can become surprisingly annoying when all transitioning to beatmap-based audio need to be seamless).

Comment (SirCmpwn) [inline osu.Game/Screens/Backgrounds/BackgroundModeBeatmap.cs]: Checking if `beatmap == Default` is the same thing as `beatmap == null`, only sillier.

The beatmap still needs to be imported into the database and written into the filesystem, which takes time. If you don't want to write it in the filesystem, the maintenance cost and complexity overhead of having the beatmap store support some kind of virtual beatmap that's not sourced from the filesystem is nuts compared to just checking if there are no beatmaps in the db.

File: osu.Game/Screens/Select/PlaySongSelect.cs [hunk 24]
```diff
@@ -174,9 +188,43 @@ private void start()
...
+                BeatmapGroup newSelection = null;
+                foreach (var beatmapGroup in carousel)
+                {
+                    var set = beatmapGroup.BeatmapSet;
+                    bool match = string.IsNullOrEmpty(search)
+                        || (set.Metadata.Artist ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.ArtistUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.Title ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1
+                        || (set.Metadata.TitleUnicode ?? "").IndexOf(search, StringComparison.InvariantCultureIgnoreCase) != -1;
+                    if (match)
```
Comment (Tom94) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: The filter will also have to work with individual difficulties; not just groups; in the future. Probably out of score for this PR, too, given that we still need to finalize how sorting e.g. by difficulty will work in the first place.

Comment (peppy) [inline osu.Game/Screens/Select/PlaySongSelect.cs]: and be regex and support a lot more things, yep :p
